source/vibe/web/rest.d

```diff
@@ -969,12 +987,13 @@ private string genClientBody(alias Func)() {
 				static if (is(PT == Json))
 					url_prefix = q{urlEncode(id.toString())~"/"};
 				else
-					url_prefix = q{urlEncode(toRestString(serializeToJson(id)))~"/"};
+					url_prefix = q{urlEncode(toRestString(id))~"/"};
 			} else static if (anySatisfy!(mixin(GenCmp!("ClientFilter", i, ParamNames[i]).Name), paramAttr)) {
 				alias paramsArgList = Filter!(mixin(GenCmp!("ClientFilter", i, ParamNames[i]).Name), UDATuple!(WebParamAttribute, Func));
-				static if (paramsArgList[0].origin == WebParamAttribute.Origin.Header)
-					param_handling_str ~= format(q{headers__["%s"] = to!string(%s);}, paramsArgList[0].field, paramsArgList[0].identifier);
-				else static if (paramsArgList[0].origin == WebParamAttribute.Origin.Query)
+				static if (paramsArgList[0].origin == WebParamAttribute.Origin.Header) {
+					static if (isNullable!PT) param_handling_str ~= format(q{ if (!%s.isNull) }, paramsArgList[0].identifier);
+					param_handling_str ~= format(q{headers__["%s"] = toRestString(%s);}, paramsArgList[0].field, paramsArgList[0].identifier);
```

Comment:
Sorry for the late reply, was quite busy lately.

> On the other hand, should anyone really be sending non-scalar values over HTTP headers? I don't think so. But if they seriously need to do it, shouldn't it be done the same way it's done everywhere else?

My personal opinion is that you should only serialize an aggregate as an aggregate if the representation allows you to. For example, it's perfectly fine to serialize an aggregate as body parameter (if we're using Json) because you're free to put whatever you want in the body, and `Content-Type` allows you to type it.
OTOH, query parameters and header parameters don't have a structure. Each parameter is essentially a single field. There are very little use cases where you want to pass an aggregate and see it serialized as Json, and in such situation, I think we should consider using `Json` parameters instead of aggregates.

I'd like to free aggregates for query and header parameters for grouping of parameters. For example, Github API has a [rate limit](https://developer.github.com/v3/#rate-limiting) which defines 3 parameters which are related to each others. Those would be good candidate for such usage. Example in code:

```D
public struct GhRate {
  @headerParam("X-RateLimit-Limit")
  uint limit;
  @headerParam("X-RateLimit-Remaining")
  uint remaining;
  @headerParam("X-RateLimit-Reset")
  time_t reset;
}

interface GhRoot {
  // Note: This uses 'out' parameters as implemented in: #1091
  Json get(out GhRate rate);
}
```

Does that sound right ?